airflow/dag/serialization.py

```diff
@@ -0,0 +1,316 @@
+            elif isinstance(x, pendulum.tz.Timezone):
+                return cls.encode(str(x.name), type_=DagTypes.TIMEZONE)
+            elif callable(x):
+                return str(get_python_source(x))
+            elif isinstance(x, set):
+                return cls.encode(
+                    [cls._serialize(v, visited_dags) for v in x], type_=DagTypes.SET)
+            elif isinstance(x, tuple):
+                return cls.encode(
+                    [cls._serialize(v, visited_dags) for v in x], type_=DagTypes.TUPLE)
+            else:
+                logging.debug('Cast type %s to str in serialization.', type(x))
+                return str(x)
+        except Exception:  # pylint: disable=broad-except
+            logging.warning('Failed to stringify.', exc_info=True)
+            return FAILED
+
+    @classmethod
+    def _deserialize(cls, encoded_x, visited_dags):  # pylint: disable=too-many-return-statements
+        """Helper function of depth first search for deserialization."""
+        try:
+            # JSON primitives (except for dict) are not encoded.
+            if cls._is_primitive(encoded_x):
+                return encoded_x
+            elif isinstance(encoded_x, list):
+                return [cls._deserialize(v, visited_dags) for v in encoded_x]
+
+            assert isinstance(encoded_x, dict)
+            var = encoded_x[Encoding.VAR]
+            type_ = encoded_x[Encoding.TYPE]
+
+            if type_ == DagTypes.DICT:
+                return {k: cls._deserialize(v, visited_dags) for k, v in var.items()}
+            elif type_ == DagTypes.DAG:
+                if isinstance(var, dict):
+                    return SerializedDAG.deserialize_dag(var, visited_dags)
+                elif isinstance(var, str) and var in visited_dags:
+                    # dag_id is stored in the serailized form for a visited DAGs.
+                    return visited_dags[var]
+                logging.warning('Invalid DAG %s in deserialization.', var)
+                return None
+            elif type_ == DagTypes.OP:
+                return SerializedOperator.deserialize_operator(var, visited_dags)
+            elif type_ == DagTypes.DATETIME:
+                return dateutil.parser.parse(var)
+            elif type_ == DagTypes.TIMEDELTA:
+                return datetime.timedelta(seconds=var)
+            elif type_ == DagTypes.TIMEZONE:
+                return pendulum.tz.timezone(name=var)
+            elif type_ == DagTypes.SET:
+                return {cls._deserialize(v, visited_dags) for v in var}
+            elif type_ == DagTypes.TUPLE:
+                return tuple([cls._deserialize(v, visited_dags) for v in var])
+            else:
+                logging.warning('Invalid type %s in deserialization.', type_)
+                return None
+        except Exception:  # pylint: disable=broad-except
+            logging.warning('Failed to deserialize %s.',
```

Comment:
Can we use the LoggingMixin here, instead of plain logging.

Reply:
Used LOG = LoggingMixin().log instead.